src/storage/src/source/sql_server/replication.rs

```diff
@@ -141,75 +151,84 @@ pub(crate) fn render<G: Scope<Timestamp = Lsn>>(
-                        })?;
+                    let snapshot_cap = data_cap_set.delayed(&snapshot_lsn);
 
-                    for (partition_idx, _) in partition_indexes {
-                        let decoder = decoder_map.get(partition_idx).expect("decoder for output");
-                        // Try to decode a row, returning a SourceError if it fails.
-                        let message = match decoder.decode(&sql_server_row, &mut mz_row, &arena) {
-                            Ok(()) => Ok(SourceMessage {
-                                key: Row::default(),
-                                value: mz_row.clone(),
-                                metadata: Row::default(),
-                            }),
-                            Err(e) => {
-                                let kind = DecodeErrorKind::Text(e.to_string().into());
-                                // TODO(sql_server2): Get the raw bytes from `tiberius`.
-                                let raw = format!("{sql_server_row:?}");
-                                Err(DataflowError::DecodeError(Box::new(DecodeError {
-                                    kind,
-                                    raw: raw.as_bytes().to_vec(),
-                                })))
-                            }
-                        };
-                        data_output
-                            .give_fueled(
-                                &snapshot_cap,
-                                ((*partition_idx, message), snapshot_lsn, Diff::ONE),
-                            )
-                            .await;
+                    snapshot_lsns.insert(Arc::clone(&table.capture_instance), Arc::new(snapshot_lsn));
+
+                    records_known = records_known.saturating_add(size);
+                    emit_stats(&stats_cap[0], records_known, records_total);
+
+                    while let Some(result) = snapshot.next().await {
+                        let sql_server_row = result.map_err(TransientError::from)?;
+
+                        records_total = records_total.saturating_add(1);
+
+                        if last_report.elapsed() > report_interval.get() {
+                            last_report = Instant::now();
+                            emit_stats(&stats_cap[0], records_known, records_total);
+                        }
+
+                        // Decode the SQL Server row into an MZ one.
+                        let mut mz_row = Row::default();
+                        let arena = RowArena::default();
+
+                        let partition_indexes = export_ids_to_snapshot.get(&table.capture_instance)
+                            .ok_or_else(|| {
+                                let msg = format!("no snapshot outputs in known capture instances [{}] for capture instance: '{}'", export_ids_to_snapshot.keys().join(","), table.capture_instance);
+                                TransientError::ProgrammingError(msg)
+                            })?;
```

Comment:
It's fine to just panic for these kin of invariants

```suggestion
                        let partition_indexes = export_ids_to_snapshot[&table.capture_instance];
```

Reply:
is there any value in having a clearer error message here? I did find it helpful when debugging it to see the list of capture instances with the capture instance we were trying to retrieve.

Reply:
You can use `.expect()` to give a message. My main point is that `TransientError` is used for things like network errors or other errors where restarting the dataflow may resolve them. In this case it's simply a bug if the capture instance doesn't exist in the map and we want to panic so that it also shows up in Sentry and we look into it asap

Reply:
ah I see, thanks. that makes a lot of sense.

Reply:
I refactored to unwrap_or_else with a panic so I could create an error message with params. hope that's okay?